src/Host/Client/Impl/Host/LocalRHostConnector.cs

```diff
@@ -31,135 +33,137 @@ internal sealed class LocalRHostConnector : IRHostConnector {
-                _log.RHostProcessExited();
-            });
-
-            var host = new RHost(name, callbacks, transportTcs.Task.Result, process.Id, cts);
-            process.Exited += delegate { host.Dispose(); };
-            return host;
-        }
+                        // Pipes are special in that a zero-length read is not an indicator of end-of-stream.
+                        // Stream.CopyTo uses a zero-length read as indicator of end-of-stream, so it cannot 
+                        // be used here. Instead, copy the data manually, using PipeStream.IsConnected to detect
+                        // when the other side has finished writing and closed the pipe.
+                        var buffer = new byte[0x1000];
+                        do {
+                            int count = await serverUriPipe.ReadAsync(buffer, 0, buffer.Length, cts.Token);
+                            serverUriData.Write(buffer, 0, count);
+                        } while (serverUriPipe.IsConnected);
+                    } catch (OperationCanceledException) {
+                        throw new RHostTimeoutException("Timed out while waiting for broker process to report its endpoint URI");
+                    }
 
+                    string serverUriStr = Encoding.UTF8.GetString(serverUriData.ToArray());
+                    Uri[] serverUri;
+                    try {
+                        serverUri = JsonConvert.DeserializeObject<Uri[]>(serverUriStr);
+                    } catch (JsonSerializationException ex) {
+                        throw new RHostDisconnectedException($"Invalid JSON for endpoint URIs received from broker ({ex.Message}): {serverUriStr}");
+                    }
+                    if (serverUri?.Length != 1) {
+                        throw new RHostDisconnectedException($"Unexpected number of endpoint URIs received from broker: {serverUriStr}");
+                    }
 
-        private static WebSocketMessageTransport CreateWebSocketMessageTransport(TaskCompletionSource<IMessageTransport> transportTcs) {
-            lock (transportTcs) {
-                if (transportTcs.Task.IsCompleted) {
-                    throw new MessageTransportException("More than one incoming connection.");
+                    Broker.BaseAddress = serverUri[0];
                 }
 
-                var transport = new WebSocketMessageTransport();
-                transportTcs.SetResult(transport);
-                return transport;
+                _brokerProcess = process;
+                DisposableBag.Add(DisposeBrokerProcess);
+            } finally {
+                if (_brokerProcess == null) {
+                    try {
+                        process?.Kill();
+                    } catch (Exception) {
+                    } finally {
+                        process?.Dispose();
+                    }
+                }
             }
         }
+
+        private void DisposeBrokerProcess() {
+            _brokerProcess?.Kill();
```

Comment:
Yep.

---
In reply to: [75045542](https://github.com/Microsoft/RTVS/pull/2173#discussion_r75045542) [](ancestors = 75045542)